File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phoneBook) {
        //1.phoneBook을 정렬한다.
        Arrays.sort(phoneBook);
        
        //2. 1중 loop를 돌면서 앞번호가 뒷번호의 접두어인지 확인한다.
        for(int i=0;i<phoneBook.length-1;i++){
            if(phoneBook[i+1].startsWith(phoneBook[i]))
                return false;
        }
        
        //3. 여가까지 오지 못했다면 접두어가 없는 것이다.
        return true;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_map={}
    
    #1. 전화번호들을 딕셔너리에 저장
    for i in range(len(phone_book)):
        phone_map[phone_book[i]]=i
    
    #2. 각 전화번호의 접두어가 딕셔너리에 존재하는지 확인
    for i in range(len(phone_book)):
        number=phone_book[i]
        for j in range(1,len(number)): #j=1부터 시작해야 접두어가 의미 있음
            prefix=number[:j]
            if prefix in phone_map:
                return False
    
    return True